core/src/main/java/org/elasticsearch/common/util/concurrent/AsyncIOProcessor.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.logging.ESLogger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.function.Consumer;
+
+/**
+ * This async IO processor allows to batch IO operations and have a single writer processing the write operations.
+ * This can be used to ensure that threads can continue with other work while the actual IO operation is still processed
+ * by a single worker. A worker in this context can be any caller of the {@link #put(Object, Consumer)} method since it will
+ * hijack a worker if nobody else is currently processing queued items. If the internal queue has reached it's capacity incoming threads
+ * might be blocked until other items are processed
+ */
+public abstract class AsyncIOProcessor<Item> {
+    private final ESLogger logger;
+    private final ArrayBlockingQueue<Tuple<Item, Consumer<Exception>>> queue;
+    private final Semaphore promiseSemaphore = new Semaphore(1);
+
+    protected AsyncIOProcessor(ESLogger logger, int queueSize) {
+        this.logger = logger;
+        this.queue = new ArrayBlockingQueue<>(queueSize);
+    }
+
+    /**
+     * Adds the given item to the queue. The listener is notified once the item is processed
+     */
+    public final void put(Item item, Consumer<Exception> listener) {
+        ensureOpen();
```

Comment:
if this throws an exception, do we need to call the listener on it to not have threads blocked forever?